features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/LinkdTopologyProvider.java

```diff
@@ -443,32 +523,109 @@ private void getCdpLinks() {
-            LinkdVertex target = (LinkdVertex)getVertex(TOPOLOGY_NAMESPACE_LINKD, targetLink.getNode().getNodeId());
-            target.getProtocolSupported().add(ProtocolSupported.CDP);
-            OnmsSnmpInterface sourceSnmpInterface = getSnmpInterface(sourceLink.getNode().getId(), sourceLink.getCdpCacheIfIndex());
-            OnmsSnmpInterface targetSnmpInterface = getSnmpInterface(targetLink.getNode().getId(), targetLink.getCdpCacheIfIndex());
+            results.add(Pair.of(sourceLink, targetLink));
+        }
+        return results;
+    }
+
+    List<Pair<CdpLink, CdpLink>> matchCdpLinksNew(final List<CdpElement> cdpElements, final List<CdpLink> allLinks) {
+
+        // 1. create lookup maps:
+        Map<Integer, CdpElement> cdpelementmap = new HashMap<Integer, CdpElement>();
+        for (CdpElement cdpelement: cdpElements) {
+            cdpelementmap.put(cdpelement.getNode().getId(), cdpelement);
+        }
+        Map<String, CdpLink> targetLinkMap = new HashMap<>();
+        for (CdpLink link : allLinks) {
+            targetLinkMap.put(createCdpLinkCacheKey(link, cdpelementmap.get(link.getNode().getId())),link);
+        }
+        Set<Integer> parsed = new HashSet<Integer>();
+
+        // 2. iterate
+        List<Pair<CdpLink, CdpLink>> results = new ArrayList<>();
+        for (CdpLink sourceLink : allLinks) {
+            if (parsed.contains(sourceLink.getId())) {
+                continue;
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("getCdpLinks: source: {} ", sourceLink.printTopology());
+            }
+            CdpElement sourceCdpElement = cdpelementmap.get(sourceLink.getNode().getId());
+
+            CdpLink targetLink = targetLinkMap.get(createCdpLinkLookupKey(sourceLink, sourceCdpElement));
+
+            if (targetLink == null) {
+                LOG.debug("getCdpLinks: cannot found target for source: '{}'", sourceLink.getId());
+                continue;
+            }
+
+            if (sourceLink.getId().equals(targetLink.getId()) || parsed.contains(targetLink.getId())) {
+                continue;
+            }
+
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("getCdpLinks: cdp: {}, target: {} ", sourceLink.getCdpCacheDevicePort(), targetLink.printTopology());
+            }
+
+            parsed.add(sourceLink.getId());
+            parsed.add(targetLink.getId());
+            results.add(Pair.of(sourceLink, targetLink));
+        }
+        return results;
+    }
+
+    private String createCdpLinkCacheKey(CdpLink targetLink, CdpElement targetElement) {
+        return  targetLink.getCdpCacheDevicePort() + "$"
+                + targetLink.getCdpInterfaceName() + "$"
+                + targetElement.getCdpGlobalDeviceId() + "$"
+                + targetLink.getCdpCacheDeviceId();
+    }
+
+    private String createCdpLinkLookupKey(CdpLink sourceLink, CdpElement sourceElement) {
```

Comment:
Can we create a new object with the necessary fields as member variables and use that object as a key rather than concatenating them to a string?

Say we have "z = x + '$' + y". When x1='x$' and y1='y', then z1='x$$y'. When x2='x' and y2='$y', then z2='x$$y'. So, z2=z1, but x1 != x2 and z1 != z2.

Reply:
sure, I replaced it by CompositeKey